### DISCUSSION TITLE
joining  multiple identical entities

### DISCUSSION THREAD

**Discussion by the1gofer** [sqlalchemy/sqlalchemy#6754] [Usage Questions] [answered]
I have a table with the following model:

```
class SimilarSources(Base):
    """table to store obit similarity so we don't have to spend time running the slow script every time"""

    __tablename__ = "similar_sources"
    id = Column(Integer, primary_key=True)
    source_one_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    source_two_id = Column(Integer, ForeignKey('sources.id'), nullable=False)
    similarity = Column(Float)

    source_one = relationship('Source', foreign_keys=[source_one_id])
    source_two = relationship('Source', foreign_keys=[source_two_id])
```

When I try to join on source, it's causing me issues.  How do I tell sqlalchamy to join both sources so I can get info out of that source model?

**Comment by zzzeek** [comment, marked as answer]
when you join to the same table twice, you need to use [SQL aliases](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#orm-entity-aliases) to disambiguate them.  So here using 1.x style it would be a query of the form:

    from sqlalchemy.orm import aliased
    s1 = aliased(Source)
    s2 = aliased(Source)
    session.query(SimilarSources).\
       join(SimilarSources.source_one.of_type(s1)).\
       join(SimilarSources.source_two.of_type(s2)).\
       filter(s1.source_name == 'name1').\
       filter(s2.source_name == 'name2')

**Reply by zzzeek** [reply]
see also https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#joining-between-aliased-targets
